virtcontainers/sandbox.go

```diff
@@ -1798,3 +1828,63 @@ func (s *Sandbox) AddDevice(info config.DeviceInfo) (api.Device, error) {
 
 	return b, nil
 }
+
+func (s *Sandbox) updateResources() error {
+	// the hypervisor.MemorySize is the amount of memory reserved for
+	// the VM and contaniners without memory limit
+
+	sumResources := specs.LinuxResources{
+		Memory: &specs.LinuxMemory{
+			Limit: new(int64),
+		},
+		CPU: &specs.LinuxCPU{
+			Period: new(uint64),
+			Quota:  new(int64),
+		},
+	}
+
+	for _, c := range s.config.Containers {
+		if m := c.Resources.Memory; m != nil && m.Limit != nil {
+			*sumResources.Memory.Limit += *m.Limit
+		}
+		if cpu := c.Resources.CPU; cpu != nil {
+			if cpu.Period != nil && cpu.Quota != nil {
+				*sumResources.CPU.Period += *cpu.Period
+				*sumResources.CPU.Quota += *cpu.Quota
+			}
+		}
+	}
+
+	sandboxVCPUs := uint32(utils.ConstraintsToVCPUs(*sumResources.CPU.Quota, *sumResources.CPU.Period))
+	sandboxVCPUs += s.hypervisor.hypervisorConfig().NumVCPUs
+
+	sandboxMemoryByte := int64(s.hypervisor.hypervisorConfig().MemorySize) << utils.MibToBytesShift
+	sandboxMemoryByte += *sumResources.Memory.Limit
+
+	// Update VCPUs
+	s.Logger().WithField("cpus-sandbox", sandboxVCPUs).Debugf("Request to hypervisor to update vCPUs")
+	oldCPUs, newCPUs, err := s.hypervisor.resizeVCPUs(sandboxVCPUs)
+	if err != nil {
+		return err
+	}
+	// The CPUs were increased, ask agent to online them
+	if oldCPUs < newCPUs {
+		vcpusAdded := newCPUs - oldCPUs
+		if err := s.agent.onlineCPUMem(vcpusAdded, true); err != nil {
+			return err
+		}
+	}
+	s.Logger().Debugf("Sandbox CPUs: %d", newCPUs)
+
+	// Update Memory
+	s.Logger().WithField("memory-sandbox-size-byte", sandboxMemoryByte).Debugf("Request to hypervisor to update memory")
+	newMemory, err := s.hypervisor.resizeMemory(uint32(sandboxMemoryByte>>utils.MibToBytesShift), s.state.GuestMemoryBlockSizeMB)
```

Comment:
As we have discuss in https://github.com/kata-containers/runtime/pull/793#discussion_r228384791 
By simply do 20 right shift to sandboxMemoryByte, we will lose memory less than 1M.
For instance, if we `docker run 128M` we will get a vm of 2G+128M memory. 
Then we `docker update 128M+4096`, we will do nothing with `MibToBytesShift` right shift here. But the memory limit cgroup of container will be 128M+4096, so it will exceed 128MB memory. 
WDYT.

Reply:
Oh, it's reasonable. Thanks for your answer.

Reply:
@jcvenegas makes sense to do that in separate PR for govmm, thanks!